File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {

    // 네트워크의 갯수 = connected component 의 갯수!
    public int solution(int n, int[][] computers) {
        // 1. 그래프 만들기
        List<Integer>[] adj = createGraph(n, computers);
        // System.out.println(Arrays.toString(adj));
        // 2. connected component 의 갯수 세기 
        int answer = countNetwork(n, adj);

        return answer;
    }

    int countNetwork(int n, List<Integer>[] adj) {
        int count = 0;
        boolean[] visited = new boolean[n];

        for (int st = 0; st < n; st++) {
            if (visited[st]) {
                continue;
            }

            bfs(st, adj, visited);
            count++;

        }

        return count;
    }

    void bfs(int st, List<Integer>[] adj, boolean[] visited) {
        visited[st] = true;
        Queue<Integer> q = new LinkedList<>();
        q.offer(st);

        while (!q.isEmpty()) {
            int sz = q.size();
            for (int i = 0; i < sz; i++) {
                int current = q.poll();

                for (int next : adj[current]) {
                    if (visited[next]) {
                        continue;
                    }

                    visited[next] = true;
                    q.offer(next);
                }

            }
        }
    } // end of bfs

    List<Integer>[] createGraph(int n, int[][] computers) {
        List<Integer>[] adj = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<>();
        }

        // i번 컴퓨터와 j번 컴퓨터가 연결되어 있으면 computers[i][j] == 1
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (i == j) { // compters[i][i]는 항상 1 
                    continue;
                }

                if (computers[i][j] == 1) {
                    adj[i].add(j);
                }
            }
        }

        return adj;
    } // end of createGraph
}

File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import deque

answer = 0

# connected component의 갯수 찾기 
def solution(n, computers):
    # 1. 그래프 만들기 -> adj, 양방향 
    
    adj = makeGraph(n, computers)
      
    global answer
    visited = [False] * n
    
    # n번 컴퓨터부터 bfs (n번 컴퓨터 부터 연결되어있는 네트워크 탐색 )
    for i in range(n):
        if visited[i]:
            continue
        bfs(adj, visited, i)
        answer += 1
    
    return answer

def bfs(adj, visited, start):
    q = deque()
    q.append(start)
    
    while len(q) != 0:
        sz = len(q)
        
        for _ in range(sz):
            current = q.pop()
            
            for i in adj[current]:
                if visited[i]:
                    continue
                
                visited[i] = True
                q.append(i)


def makeGraph(n, computers):
    adj = [[] for _ in range(n)]
    for i in range(len(computers)):
        for j in range(len(computers[i])):
            if i == j:
                continue
            if computers[i][j] == 1:
                adj[i].append(j)
    return adj